### PR TITLE
Add Index to LSM Tree System

### PR DESCRIPTION
**Pull request by mmfrenkel**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -13,16 +14,16 @@ LDFLAGS =  -g
...
 	$(CC) $(LDFLAGS) $^ $(LDLIBS) -o $@
 
 $(OBJ_DIR)/%.o: $(SRC_DIR)/%.c | $(OBJ_DIR)
 	$(CC) $(CPPFLAGS) $(CFLAGS) -c $< -o $@
 
-$(BIN_DIR) $(OBJ_DIR):
+$(BIN_DIR) $(OBJ_DIR) $(LOG_DIR):
 	mkdir -p $@
 
 clean:
```
Comment (andyg7) [inline Makefile:28]: I'd create a new target for deleting the `logs` directory. Just re-compiling the program wouldn't seem to warrant deleting the database.

File: src/index.c [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+		return NULL;
+	}
+
+	// create contents and set all pointers to null
+	Entry **contents = (Entry **) malloc(size * sizeof(Entry*));
+	for (int i = 0; i < size; i++) {
+		*(contents + i) = NULL;
+	}
+
+	if (contents == NULL) {
```
Comment (andyg7) [inline src/index.c:17]: This check should be above the for loop.

File: src/lsm_tree.c [hunk 3]
```diff
@@ -51,10 +48,20 @@ LSM_Tree* init_lsm_tree() {
 		return NULL;
 	}
 
+	Index *index = init_index(INDEX_SIZE);
+	if (wal == NULL) {
```
Comment (andyg7) [inline src/lsm_tree.c]: Should this be:
```
if (index == NULL) {
...
```
?

File: src/index.c [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+	return index;
+}
+
+/* Inserts a new entry into hash table; where there are
+ * collisions, this program uses chaining. */
+int index_insert(Index *index, int key, char *value) {
+
+	int position = hash(key, index->size);
+
+	Entry *new_entry = (Entry*) malloc(sizeof(Entry));
```
Comment (andyg7) [inline src/index.c]: Didn't we already allocate memory for the `Entry` when the index was initialized?

File: src/index.c [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+	new_entry->next = NULL;
+
+	// see if the key is already in the hash map
+	Entry *in_slot = *(index->contents + position);
+	Entry *trav = in_slot;
+	if (trav) {
+		while (trav) {
+			if (trav->key == key) {
+				trav->value = value;
+				return 0;
```
Comment (andyg7) [inline src/index.c:46]: In this case, do we need to free the memory we allocated above on line 39? Maybe you could find the location first and only then allocate memory if needed. If we already have an `Entry` for the key, it looks like we can just re-use the allocated `Entry`

File: src/index.c [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+
+			free(entries);
+			return 0;
+		}
+		trail = entries;
+		entries = entries->next;
+	}
+
+	// if we get to here, key is not in table so
+	printf("Key to remove not in index");
```
Comment (andyg7) [inline src/index.c]: np; add `key` to print statement

File: src/index.c [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+			return entries->value;
+		}
+		entries = entries->next;
+	}
+	// if we get to here, key is not in table
+	printf("Key is not currently in index.\n");
+	return NULL;
+}
+
+int index_remove(Index *index, int key) {
```
Comment (andyg7) [inline src/index.c:92]: Just for my understanding, when is `index_remove` used?

File: README.md [hunk 8]
```diff
@@ -1,6 +1,6 @@
 # Mini LSM Tree Database System
 
-This C project represents a simplified version of a log structed merge (LSM) tree database system. 
+This C project represents a simplified version of a log-structed merge (LSM) tree database system.
```
Comment (andyg7) [inline README.md]: I'd just say `This C project is a simplified...`

File: src/index.c [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+		*(contents + i) = NULL;
+	}
+
+	if (contents == NULL) {
+		printf("Failed to allocate memory for index contents.\n");
+		free(index);
+		return NULL;
+	}
+
+	index->size = size;
```
Comment (andyg7) [inline src/index.c]: np; I feel "capacity" might be a clearer name than "size".

File: src/segment.c [hunk 10]
```diff
@@ -74,3 +87,173 @@ int delete_segment(char *filename) {
...
+}
+
+/* Used to perform compaction step of two segment files. This method is
+ * necessary for cleaning up old segment files and keeping read I/O from
+ * getting out of control. Merges old segments together into new segments.*/
+static int merge_segments(char *filename_a, char *filename_b,
+						  char *new_segment_name, int line_size, char *tombstone) {
+
+	// open files for segments of interest
+	FILE *seg_ptr_a = fopen(filename_a, "r");
```
Comment (andyg7) [inline src/segment.c]: Error handling if `fopen` fails?

**Review by andyg7 — COMMENTED** [review]
Nice clean code overall.

File: Makefile (hunk 1 above)
Comment (mmfrenkel) [reply Makefile:28]: this is a good idea!

File: src/index.c (hunk 2 above)
Comment (mmfrenkel) [reply src/index.c:17]: nice catch!

File: src/lsm_tree.c (hunk 3 above)
Comment (mmfrenkel) [reply src/lsm_tree.c]: 🐛

File: src/index.c (hunk 4 above)
Comment (mmfrenkel) [reply src/index.c]: No, we allocated memory for a list of pointers to `Entry` structs.

File: src/index.c (hunk 5 above)
Comment (mmfrenkel) [reply src/index.c:46]: ya make sense, doing that now.

File: src/index.c (hunk 6 above)
Comment (mmfrenkel) [reply src/index.c]: 👍

File: src/index.c (hunk 7 above)
Comment (mmfrenkel) [reply src/index.c:92]: yea that's a good question, I think maybe I didn't fully implement this actually yet haha it's supposed to be called when a deleted node goes to a segment -- I just updated this.. whoops!

File: README.md (hunk 8 above)
Comment (mmfrenkel) [reply README.md]: 👍

File: src/index.c (hunk 9 above)
Comment (mmfrenkel) [reply src/index.c]: 👍

File: src/segment.c (hunk 10 above)
Comment (mmfrenkel) [reply src/segment.c]: good call I didn't do anything about these.

File: src/index.c (hunk 5 above)
Comment (andyg7) [reply src/index.c:46]: Thanks.

**Pull request merged**